File: C#/POxO/POxOSerializerUtil.cs
﻿/*
 * Copyright 2014 Giuseppe Gerla. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Reflection;

namespace POxO
{
    public class POxOSerializerUtil
    {
        private Dictionary<Type, ConstructorInfo> typeConstructorMap;

        private Dictionary<String, Type> classForName;
        private Dictionary<Type, String> nameForClass;

        private Dictionary<Type, GenericClassSerializer> serializerForClass;

        public POxOSerializerUtil()
        {
            classForName = new Dictionary<String, Type>();
            nameForClass = new Dictionary<Type, String>();
            typeConstructorMap = new Dictionary<Type, ConstructorInfo>();
            serializerForClass = new Dictionary<Type, GenericClassSerializer>();
            initializePrimitiveType();
        }

        private void initializePrimitiveType()
        {
            nameForClass.Add(typeof(Int32), "int");
            nameForClass.Add(typeof(Int64), "long");
            nameForClass.Add(typeof(Int16), "short");
            nameForClass.Add(typeof(Double), "double");
            nameForClass.Add(typeof(Single), "float");
            nameForClass.Add(typeof(Boolean), "bool");
            nameForClass.Add(typeof(SByte), "byte");
            nameForClass.Add(typeof(Char), "char");
            nameForClass.Add(typeof(String), "string");
            nameForClass.Add(typeof(DateTime), "date");
            nameForClass.Add(typeof(Enum), "enum");
            
            classForName.Add("int", typeof(Int32));
            classForName.Add("long", typeof(Int64));
            classForName.Add("short", typeof(Int16));
            classForName.Add("double", typeof(Double));
            classForName.Add("float", typeof(float));
            classForName.Add("bool", typeof(Boolean));
            classForName.Add("byte", typeof(SByte));
            classForName.Add("char", typeof(Char));
            classForName.Add("string", typeof(String));
            classForName.Add("date", typeof(DateTime));
            classForName.Add("enum", typeof(Enum));
            classForName.Add("list", typeof(IList<>));
            classForName.Add("set", typeof(ISet<>));
            classForName.Add("map", typeof(IDictionary<,>));

            serializerForClass.Add(typeof(Object), new ObjectSerializer(this));
            serializerForClass.Add(typeof(int), new IntegerSerializer(typeof(int)));
            serializerForClass.Add(typeof(int?), new IntegerSerializer(typeof(int?)));
            serializerForClass.Add(typeof(long), new LongSerializer(typeof(long)));
            serializerForClass.Add(typeof(long?), new LongSerializer(typeof(long?)));
            serializerForClass.Add(typeof(short), new ShortSerializer(typeof(short)));
            serializerForClass.Add(typeof(short?), new ShortSerializer(typeof(short?)));
            serializerForClass.Add(typeof(double), new DoubleSerializer(typeof(double)));
            serializerForClass.Add(typeof(double?), new DoubleSerializer(typeof(double?)));
            serializerForClass.Add(typeof(float), new FloatSerializer(typeof(float)));
            serializerForClass.Add(typeof(float?), new FloatSerializer(typeof(float?)));
            serializerForClass.Add(typeof(bool), new BooleanSerializer(typeof(bool)));
            serializerForClass.Add(typeof(bool?), new BooleanSerializer(typeof(bool?)));
            serializerForClass.Add(typeof(sbyte), new ByteSerializer(typeof(sbyte)));
            serializerForClass.Add(typeof(sbyte?), new ByteSerializer(typeof(sbyte?)));
            serializerForClass.Add(typeof(char), new CharSerializer(typeof(char)));
            serializerForClass.Add(typeof(char?), new CharSerializer(typeof(char?)));
            serializerForClass.Add(typeof(string), new StringSerializer());
            serializerForClass.Add(typeof(DateTime), new DateSerializer());
            serializerForClass.Add(typeof(Enum), new EnumSerializer(typeof(Enum)));
        }

        public Object createNewInstance(Type clazz)
        {
            Object ret = null;
            if (!typeConstructorMap.ContainsKey(clazz))
            {
                try
                {
                    ConstructorInfo cInfo = clazz.GetConstructor(BindingFlags.NonPublic | BindingFlags.CreateInstance | BindingFlags.Instance | BindingFlags.Public,
                                    null, new Type[] { }, null);
                    typeConstructorMap.Add(clazz, cInfo);
                    ret = cInfo.Invoke(new object[] { });
                }
                catch
                {
                    ret = null;
                }
            }
            else
            {
                ConstructorInfo cInfo = typeConstructorMap[clazz];
                ret = cInfo.Invoke(new object[] { });
            }


            return ret;
        }

        public GenericClassSerializer GetTypeSerializer(Type fieldType)
        {
            GenericClassSerializer ret = null;
            if (serializerForClass.ContainsKey(fieldType))
            {
                ret = serializerForClass[fieldType];
            }
            else
            {
                if (typeof(Enum).IsAssignableFrom(fieldType))
                {
                    ret = new EnumSerializer(fieldType);
                    serializerForClass.Add(fieldType, ret);
                }
                else
                {
                    if (fieldType.IsGenericType)
                    {
                        if ((fieldType.GetGenericTypeDefinition() == typeof(IList<>)) || (fieldType.GetGenericTypeDefinition() == typeof(List<>)))
                        {
                            POxOSerializerClassPair pair = new POxOSerializerClassPair();
                            pair.setGenericClass(typeof(Object));
                            pair.setSerializer(serializerForClass[typeof(Object)]);
                            ret = new ListSerializer(pair);
                            serializerForClass.Add(fieldType, ret);
                        }
                        else
                        {
                            if ((fieldType.GetGenericTypeDefinition() == typeof(ISet<>)) || (fieldType.GetGenericTypeDefinition() == typeof(HashSet<>)))
                            {
                                Type genType = fieldType.GetGenericArguments()[0];
                                POxOSerializerClassPair pair = new POxOSerializerClassPair();
                                pair.setGenericClass(typeof(Object));
                                pair.setSerializer(serializerForClass[typeof(Object)]);
                                ret = new SetSerializer(pair);
                                serializerForClass.Add(fieldType, ret);
                            }
                            else
                            {
                                if ((fieldType.GetGenericTypeDefinition() == typeof(IDictionary<,>)) || (fieldType.GetGenericTypeDefinition() == typeof(Dictionary<,>)))
                                {
                                    POxOSerializerClassPair keyPair = new POxOSerializerClassPair();
                                    keyPair.setGenericClass(typeof(Object));
                                    keyPair.setSerializer(serializerForClass[typeof(Object)]);
                                    POxOSerializerClassPair valuePair = new POxOSerializerClassPair();
                                    valuePair.setGenericClass(typeof(Object));
                                    valuePair.setSerializer(serializerForClass[typeof(Object)]);
                                    ret = new MapSerializer(keyPair, valuePair);
                                    serializerForClass.Add(fieldType, ret);
                                }
                            }
                        }
                    }
                    else
                    {
                        ret = serializerForClass[typeof(Object)];
                    }
                }
            }

            return ret;
        }

        public GenericClassSerializer GetFieldSerializer(FieldInfo field)
        {
            GenericClassSerializer ret = null;
            Type fieldType = field.FieldType;
            if (fieldType.IsGenericType &&
                ((fieldType.GetGenericTypeDefinition() == typeof(IList<>)) || (fieldType.GetGenericTypeDefinition() == typeof(List<>))
            || (fieldType.GetGenericTypeDefinition() == typeof(ISet<>)) || (fieldType.GetGenericTypeDefinition() == typeof(HashSet<>))
            || (fieldType.GetGenericTypeDefinition() == typeof(IDictionary<,>)) || (fieldType.GetGenericTypeDefinition() == typeof(Dictionary<,>))))
            {
                POxOSerializerClassPair pair = new POxOSerializerClassPair();
                recursiveFindSerializer(fieldType, pair);
                ret = pair.getSerializer();
            }
            else
            {
                ret = GetTypeSerializer(fieldType);
            }
            return ret;
        }


        private void recursiveFindSerializer(Type genericType, POxOSerializerClassPair pair) {
            if (!genericType.IsGenericType)
            {
                pair.setGenericClass(genericType);
                pair.setSerializer(GetTypeSerializer(genericType));
            }
            else
            {
                Type[] genericTypes = genericType.GetGenericArguments();
                POxOSerializerClassPair[] nestedPairs = new POxOSerializerClassPair[genericTypes.Length];
                for (int i = 0; i < genericTypes.Length; i++)
                {
                    Type actualType = genericTypes[i];
                    POxOSerializerClassPair nestedPair = new POxOSerializerClassPair();
                    recursiveFindSerializer(actualType, nestedPair);
                    nestedPairs[i] = nestedPair;
                }
                if ((genericType.GetGenericTypeDefinition() == typeof(IList<>)) || (genericType.GetGenericTypeDefinition() == typeof(List<>)))
                {
                    pair.setGenericClass(genericType);
                    pair.setSerializer(new ListSerializer(nestedPairs[0]));
                }
                else
                {
                    if ((genericType.GetGenericTypeDefinition() == typeof(ISet<>)) || (genericType.GetGenericTypeDefinition() == typeof(HashSet<>)))
                    {
                        pair.setGenericClass(genericType);
                        pair.setSerializer(new SetSerializer(nestedPairs[0]));
                    }
                    else
                    {
                        if ((genericType.GetGenericTypeDefinition() == typeof(IDictionary<,>)) || (genericType.GetGenericTypeDefinition() == typeof(Dictionary<,>)))
                        {
                            pair.setGenericClass(genericType);
                            pair.setSerializer(new MapSerializer(nestedPairs[0],
                                            nestedPairs[1]));
                        }
                        else
                        {
                            if ((genericType.GetGenericTypeDefinition() == typeof(Nullable<>)))
                            {
                                pair.setGenericClass(genericType);
                                pair.setSerializer(new NullableSerializer(nestedPairs[0]));
                            }
                        }
                    }
                }
            }
        }

        public Type getClassFromName(String className)
        {
            Type type = null;
            if (!classForName.ContainsKey(className))
            {
                type = LoadTypeFromName(className);
                classForName.Add(className, type);
                nameForClass.Add(type, className);
            }
            else
            {
                type = classForName[className];
            }

            return type;
        }

        public String getNameFromClass(Type type)
        {
            String name = null;
            if(type.IsGenericType)
            {
                if ((type.GetGenericTypeDefinition() == typeof(IList<>)) || (type.GetGenericTypeDefinition() == typeof(List<>)))
                {
                    name = "list";
                }
                else
                {
                    if ((type.GetGenericTypeDefinition() == typeof(ISet<>)) || (type.GetGenericTypeDefinition() == typeof(HashSet<>)))
                    {
                        name = "set";
                    }
                    else
                    {
                        if ((type.GetGenericTypeDefinition() == typeof(IDictionary<,>)) || (type.GetGenericTypeDefinition() == typeof(Dictionary<,>)))
                        {
                            name = "map";
                        }
                    }
                }
            }
            else
            {
                if (!nameForClass.ContainsKey(type))
                {
                    name = type.FullName;
                    nameForClass.Add(type, name);
                    classForName.Add(name, type);
                }
                else
                {
                    name = nameForClass[type];
                }
            }

            return name;
        }

        private Type LoadTypeFromName(String typeName)
        {
            Type ret = null;


            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
            // the trouble is that we don't know which assembly the class is defined in,
            // because we are using the "Web Site" model in Visual Studio that compiles
            // them on the fly into assemblies with random names
            // -> however, we do know that the assembly will be named App_Web_*
            // (http://msdn.microsoft.com/en-us/magazine/cc163496.aspx)
            foreach (Assembly assembly in assemblies)
            {
                Type t = assembly.GetType(typeName);
                if (t != null)
                {
                    ret = t;
                    break;
                }
            }

            if (ret == null)
                throw new ArgumentException("Unknown type to deserialize " + typeName);

            return ret;
        }
    }
}

File: Java/src/main/java/com/github/ggerla/poxoserializer/POxOSerializerUtil.java
/*
 * Copyright 2014 Giuseppe Gerla. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.ggerla.poxoserializer;

import com.github.ggerla.poxoserializer.exception.POxOSerializerException;
import com.github.ggerla.poxoserializer.serializers.BooleanSerializer;
import com.github.ggerla.poxoserializer.serializers.ByteSerializer;
import com.github.ggerla.poxoserializer.serializers.CharSerializer;
import com.github.ggerla.poxoserializer.serializers.DateSerializer;
import com.github.ggerla.poxoserializer.serializers.DoubleSerializer;
import com.github.ggerla.poxoserializer.serializers.EnumSerializer;
import com.github.ggerla.poxoserializer.serializers.FloatSerializer;
import com.github.ggerla.poxoserializer.serializers.GenericClassSerializer;
import com.github.ggerla.poxoserializer.serializers.IntegerSerializer;
import com.github.ggerla.poxoserializer.serializers.ListSerializer;
import com.github.ggerla.poxoserializer.serializers.LongSerializer;
import com.github.ggerla.poxoserializer.serializers.MapSerializer;
import com.github.ggerla.poxoserializer.serializers.ObjectSerializer;
import com.github.ggerla.poxoserializer.serializers.ShortSerializer;
import com.github.ggerla.poxoserializer.serializers.StringSerializer;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.GenericArrayType;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class POxOSerializerUtil {
	private Map<String, Constructor<?>> constructrForClass;

	private Map<String, Class<?>> classForName;
	private Map<Class<?>, String> nameForClass;

	private Map<Class<?>, GenericClassSerializer> serializerForClass;

	private ClassLoader classLoader;

	public POxOSerializerUtil() {
		constructrForClass = new TreeMap<String, Constructor<?>>();
		classForName = new TreeMap<String, Class<?>>();
		nameForClass = new HashMap<Class<?>, String>();
		serializerForClass = new HashMap<Class<?>, GenericClassSerializer>();
		classLoader = this.getClass().getClassLoader();
		initializePrimitiveType();
	}

	private void initializePrimitiveType() {
		serializerForClass.put(Object.class, new ObjectSerializer(this));

		nameForClass.put(Integer.class, "int");
		serializerForClass.put(Integer.class, new IntegerSerializer(
				Integer.class));

		nameForClass.put(int.class, "int");
		serializerForClass.put(int.class, new IntegerSerializer(int.class));

		nameForClass.put(Long.class, "long");
		serializerForClass.put(Long.class, new LongSerializer(Long.class));

		nameForClass.put(long.class, "long");
		serializerForClass.put(long.class, new LongSerializer(long.class));

		nameForClass.put(Short.class, "short");
		serializerForClass.put(Short.class, new ShortSerializer(Short.class));

		nameForClass.put(short.class, "short");
		serializerForClass.put(short.class, new ShortSerializer(short.class));

		nameForClass.put(Double.class, "double");
		serializerForClass
				.put(Double.class, new DoubleSerializer(Double.class));

		nameForClass.put(double.class, "double");
		serializerForClass
				.put(double.class, new DoubleSerializer(double.class));

		nameForClass.put(Float.class, "float");
		serializerForClass.put(Float.class, new FloatSerializer(Float.class));

		nameForClass.put(float.class, "float");
		serializerForClass.put(float.class, new FloatSerializer(float.class));

		nameForClass.put(Boolean.class, "bool");
		serializerForClass.put(Boolean.class, new BooleanSerializer(
				Boolean.class));

		nameForClass.put(boolean.class, "bool");
		serializerForClass.put(boolean.class, new BooleanSerializer(
				boolean.class));

		nameForClass.put(Byte.class, "byte");
		serializerForClass.put(Byte.class, new ByteSerializer(Byte.class));

		nameForClass.put(byte.class, "byte");
		serializerForClass.put(byte.class, new ByteSerializer(byte.class));

		nameForClass.put(Character.class, "char");
		serializerForClass.put(Character.class, new CharSerializer(
				Character.class));

		nameForClass.put(char.class, "char");
		serializerForClass.put(char.class, new CharSerializer(char.class));

		nameForClass.put(String.class, "string");
		serializerForClass.put(String.class, new StringSerializer());

		nameForClass.put(Date.class, "date");
		serializerForClass.put(Date.class, new DateSerializer());

		nameForClass.put(Enum.class, "enum");
		serializerForClass.put(Enum.class, new EnumSerializer(Enum.class));

		nameForClass.put(List.class, "list");
		nameForClass.put(Map.class, "map");

		classForName.put("int", Integer.class);
		classForName.put("long", Long.class);
		classForName.put("short", Short.class);
		classForName.put("double", Double.class);
		classForName.put("float", Float.class);
		classForName.put("bool", Boolean.class);
		classForName.put("byte", Byte.class);
		classForName.put("char", Character.class);
		classForName.put("string", String.class);
		classForName.put("date", Date.class);
		classForName.put("enum", Enum.class);
		classForName.put("list", List.class);
		classForName.put("map", Map.class);
	}

	@SuppressWarnings("unchecked")
	public <T> T createNewInstance(Class<? extends T> genericClass)
			throws InstantiationException, IllegalAccessException,
			IllegalArgumentException, InvocationTargetException {

		Constructor<?> ctor = constructrForClass.get(genericClass.getName());
		if (ctor == null) {
			Constructor<?>[] ctors = genericClass.getDeclaredConstructors();
			for (int i = 0; i < ctors.length; i++) {
				ctor = ctors[i];
				if (ctor.getGenericParameterTypes().length == 0)
					break;
			}

			ctor.setAccessible(true);
			constructrForClass.put(genericClass.getName(), ctor);
		}
		T ret = (T) ctor.newInstance();

		return ret;
	}

	public GenericClassSerializer getTypeSerializer(Class<?> fieldType)
			throws POxOSerializerException {
		GenericClassSerializer ret = serializerForClass.get(fieldType);
		if (ret == null) {
			if (Enum.class.isAssignableFrom(fieldType)) {
				ret = new EnumSerializer(fieldType);
				serializerForClass.put(fieldType, ret);
			} else {
				ret = serializerForClass.get(Object.class);
			}
		}

		return ret;
	}

	public GenericClassSerializer getFieldSerializer(Field field)
			throws POxOSerializerException {
		GenericClassSerializer ret = null;
		Class<?> fieldType = field.getType();
		if (List.class.isAssignableFrom(fieldType)) {
			POxOSerializerClassPair pair = new POxOSerializerClassPair();
			recirsiveFindSerializer(field.getGenericType(), pair);
			ret = pair.getSerializer();
		} else if (Map.class.isAssignableFrom(fieldType)) {
			POxOSerializerClassPair pair = new POxOSerializerClassPair();
			recirsiveFindSerializer(field.getGenericType(), pair);
			ret = pair.getSerializer();
		} else {
			ret = getTypeSerializer(fieldType);
		}
		return ret;
	}

	private void recirsiveFindSerializer(Type genericType,
			POxOSerializerClassPair pair) throws POxOSerializerException {
		if (genericType instanceof GenericArrayType) {
			Type componentType = ((GenericArrayType) genericType)
					.getGenericComponentType();
			if (componentType instanceof Class<?>) {
				pair.setGenericClass((Class<?>) componentType);
				pair.setSerializer(new ListSerializer((Class<?>) componentType,
						getTypeSerializer((Class<?>) componentType)));
				return;
			} else {
				POxOSerializerClassPair nestedPair = new POxOSerializerClassPair();
				recirsiveFindSerializer(componentType, nestedPair);
				pair.setGenericClass(nestedPair.getGenericClass());
				pair.setSerializer(new ListSerializer(nestedPair
						.getGenericClass(), nestedPair.getSerializer()));
				return;
			}
		}
		if (genericType instanceof ParameterizedType) {
			Type[] actualTypes = ((ParameterizedType) genericType)
					.getActualTypeArguments();
			Class<?> genericClass = (Class<?>) ((ParameterizedType) genericType)
					.getRawType();
			if (Map.class.isAssignableFrom(genericClass)) {
				POxOSerializerClassPair[] serializers = new POxOSerializerClassPair[actualTypes.length];
				for (int i = 0, n = actualTypes.length; i < n; i++) {
					Type actualType = actualTypes[i];

					if (actualType instanceof Class<?>) {
						POxOSerializerClassPair nestedPair = new POxOSerializerClassPair();
						nestedPair.setGenericClass((Class<?>) actualType);
						nestedPair
								.setSerializer(getTypeSerializer((Class<?>) actualType));
						serializers[i] = nestedPair;
					} else if (actualType instanceof ParameterizedType) {
						POxOSerializerClassPair nestedPair = new POxOSerializerClassPair();
						recirsiveFindSerializer(actualType, nestedPair);
						serializers[i] = nestedPair;
					} else
						continue;
				}
				pair.setGenericClass((Class<?>) ((ParameterizedType) genericType)
						.getRawType());
				pair.setSerializer(new MapSerializer(serializers[0]
						.getGenericClass(), serializers[1].getGenericClass(),
						serializers[0].getSerializer(), serializers[1]
								.getSerializer()));
			} else if (List.class.isAssignableFrom(genericClass)) {
				for (int i = 0, n = actualTypes.length; i < n; i++) {
					Type actualType = actualTypes[i];

					if (actualType instanceof Class<?>) {
						pair.setGenericClass((Class<?>) ((ParameterizedType) genericType)
								.getRawType());
						pair.setSerializer(new ListSerializer(
								(Class<?>) actualType,
								getTypeSerializer((Class<?>) actualType)));
					} else if (actualType instanceof ParameterizedType) {
						POxOSerializerClassPair nestedPair = new POxOSerializerClassPair();
						recirsiveFindSerializer(actualType, nestedPair);
						pair.setGenericClass((Class<?>) ((ParameterizedType) genericType)
								.getRawType());
						pair.setSerializer(new ListSerializer(nestedPair
								.getGenericClass(), nestedPair.getSerializer()));
					} else
						continue;
				}
			}
		}
	}

	public Class<?> getClassFromName(String className)
			throws POxOSerializerException {
		Class<?> type = classForName.get(className);
		if (type == null) {
			try {
				type = classLoader.loadClass(className);
			} catch (ClassNotFoundException e) {
				throw new POxOSerializerException("Error during loading class "
						+ className);
			}
			classForName.put(className, type);
			nameForClass.put(type, className);
		}

		return type;
	}

	public String getNameFromClass(Class<?> type) {
		String name = nameForClass.get(type);
		if (name == null) {
			name = type.getName();
			nameForClass.put(type, name);
			classForName.put(name, type);
		}
		return name;
	}

	public void setClassLoader(ClassLoader classLoader) {
		this.classLoader = classLoader;
	}
}
